pkg/sse/v1alpha/sse.go

```diff
@@ -0,0 +1,235 @@
+}
+
+// Selftest is a handler to just return 204.
+func (h SSEHandler) Selftest(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	responses.RespondWithError(w, http.StatusNoContent, "")
+}
+
+// Subscribe subscribes to an ETOS suite runner instance and gets logs and events from it and
+// writes them to a channel.
+func (h SSEHandler) Subscribe(ch chan<- events.Event, ctx context.Context, counter int, identifier string, url string) {
+	defer close(ch)
+
+	tick := time.NewTicker(1 * time.Second)
+	defer tick.Stop()
+	ping := time.NewTicker(15 * time.Second)
+	defer ping.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			h.logger.Info("Client lost, closing subscriber")
+			return
+		case <-ping.C:
+			ch <- events.Event{Event: "ping"}
+		case <-tick.C:
+			if h.kube.IsFinished(ctx, identifier) {
+				h.logger.Info("ESR finished, shutting down")
+				ch <- events.Event{Event: "shutdown"}
+				return
+			}
+			messages, err := GetFrom(ctx, url)
+			if err != nil {
+				h.logger.Warning(err.Error())
+				continue
+			}
+			// SSE starts ad ID 1, but slices start with index 0, so we do -1 here.
+			if len(messages) >= counter-1 {
+				for _, message := range messages[counter-1:] {
+					event := events.Event{
+						Event: "message",
+						ID:    counter,
+						Data:  message,
+					}
+					ch <- event
+					counter++
+				}
+			}
+		}
+	}
+}
+
+// GetFrom gets all events from an ESR instance
+func GetFrom(ctx context.Context, url string) ([]string, error) {
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := http.DefaultClient.Do(request)
+	if err != nil {
+		return nil, err
+	}
+
+	var messages []string
+	scanner := bufio.NewScanner(response.Body)
+	defer response.Body.Close()
+	for scanner.Scan() {
+		messages = append(messages, scanner.Text())
+	}
+	return messages, nil
+}
+
+// url finds the url of an ESR instance.
+func (h SSEHandler) url(ctx context.Context, identifier string) (string, error) {
+	ip, err := h.kube.LogListenerIP(ctx, identifier)
+	if err != nil {
+		return "", err
+	}
+	if ip == "" {
+		return "", errors.New("No IP from ESR yet")
+	}
+	return fmt.Sprintf("http://%s:8000/log", ip), nil
+}
+
+// forceKillConnection hijacks the underlying TCP connection between the client and server
+// and stops it forcefully. This will cause a panic in the goroutine running the connection
+// but this is the only way for us to be compatible with the old ETOS SSE implementation.
+func forceKillConnection(w http.ResponseWriter) {
+	hijack, ok := w.(http.Hijacker)
+	if !ok {
+		responses.RespondWithError(w, http.StatusInternalServerError, http.StatusText(http.StatusInternalServerError))
+		return
+	}
+	connection, _, err := hijack.Hijack()
+	if err != nil {
+		responses.RespondWithError(w, http.StatusInternalServerError, http.StatusText(http.StatusInternalServerError))
+		return
+	}
+	connection.Close()
+}
```

Comment:
Perhaps add some logging about the abuse happening here. So we don't get confused by potential panic logs.